v2/apiserver/internal/core/events.go

```diff
@@ -537,35 +537,59 @@ func (e *eventsService) CancelMany(
 		)
 	}
 
-	events, err := e.eventsStore.CancelMany(ctx, selector)
+	eventCh, errCh, err := e.eventsStore.CancelMany(ctx, selector)
 	if err != nil {
 		return result, errors.Wrap(err, "error canceling events in store")
 	}
 
-	result.Count = int64(len(events.Items))
-
-	// TODO: This could take a while, so we don't do it synchronously. But what if
-	// the process dies while this is in-progress? Can we find a quicker, more
-	// efficient way to do this?
-	go func() {
-		for _, event := range events.Items {
-			if err := e.substrate.DeleteWorkerAndJobs(
-				context.Background(), // Deliberately not using request context
-				project,
-				event,
-			); err != nil {
-				log.Println(
-					errors.Wrapf(
-						err,
-						"error deleting event %q worker and jobs from the substrate",
-						event.ID,
-					),
-				)
+	// When an event comes in via the corresponding channel,
+	// kick off a go routine to execute cleanup.
+	// Otherwise, handle errors and/or finish and return once all
+	// events have come through.
+	for {
+		select {
+		case event, ok := <-eventCh:
+			if ok {
+				result.Count++
+				// TODO: This could take a while, so we don't do it synchronously. But
+				// what if the process dies while this is in-progress? Can we find a
+				// quicker, more efficient way to do this?
+				go func() {
+					if err := e.substrate.DeleteWorkerAndJobs(
```

Comment:
I was about to propose doing this conditionally-- only if the event's worker's current phase is `ABORTED`. If the current phase were `CANCELED`, that worker never made it past `PENDING` and cannot possibly have any pods associated with it...

Then I remembered that there's still a secret associated with every pending event that still needs to be removed upon cancellation.

But what we could do to speed up the whole process is do that phase in the substrate's `DeleteWorkerAndJobs()` function. If canceled, only delete secrets associated with the event. If aborted, delete associated secrets _and_ pods.

Reply:
I'm fine with that being a in a follow-up, but it could also make sense to tack it onto this PR...

Dealer's choice.

Reply:
Good call; it appeared simple enough to include.  Does https://github.com/brigadecore/brigade/pull/1272/commits/2daed15b0d954d210e74b090fcfd08c05ae2d0d6 look good?